Optimize phpdoc

voku opened the pull request:
helpful for e.g. PhpStorm but also for PHPStan itself

-> thx @ PhpStorm + Php Inspections (EA Ultimate)

ondrejmirtes commented:
Hi, I don't see the point - PHPStan is able to infer many cases even without the phpDoc, and some of those changes are even wrong, for example NodeScopeResolver, line 781 (`						/** @var StaticPropertyFetch $unsetVar */`). Additionally, PHPStan is able to make use of `@var`s only  above the assignment of the same variable.

These kinds of changes in the whole codebase I don't like to do very often, they should be informed and supported by PHPStan itself, ideally by a new rule (both internal and external), influencing the CI build as well.

voku commented:
Hi, I understand your point and a rule for adding ```/** @var``` comments for e.g. PhpStorm is a very good idea and sounds like a "PHPStan Pro"-feature. :)

I added this things only for the reason that my IDE complains about it and so I did not see if it's a real error or only some random garbage from the PhpStorm Inspection. 

---

PS: I also wanted to give you some feedback (I use it for a legacy application), maybe it can help you for "PHPStan Pro" -> here a mix of my current setup and some ideas :) 

#### 1. Old PHP support for "enterprise" applications
Our project (and mostly especially the servers) has moved to PHP 7.0 now, so most developers also use PHP 7.0 for there local environment and so I needed to add a wrapper (something like this: https://gist.github.com/voku/c78c73c0c34828462369e2d242622c93) to use new stuff if possible and old stuff when needed. So I added "phpstan_php7_1.phar" (0.10.*) and also "phpstan_php7_0.phar" (0.9.*) into the thirdparty directory. I know we should use docker and everybody should have the same environment :-/ but for now the base setup is very simple and nobody really wants to use docker for now. So running PHPStan in the cloud (as service) sounds good, but it need to run also on the developers machine.

=> I think many projects for the next years (e.g. Ubuntu 16.04 LTS - supports PHP 7.0 unit the year 2021) will use PHP 7.0 :( and I know there are no nullable return types and no more official support etc., but the dev-op will not install PHP 7.1, because some guys from the linux distributions (http://phpversions.info/operating-systems/) will try to backport the fixes into PHP 7.0. :( Another example is Red Hat Enterprise Linux 7.x, it will support PHP 5.x until the year 2027!?!? :face_with_head_bandage: So maybe some more documentation on how to use it with legacy application and how "PHPStan Pro" can help you here, is a good idea. For example I did not know that the ```/** @var``` thingy (that is used often in legacy application where not everything has a return type) works only above the assignment of the same variable. I now added the following text in our small code quality tool documentation: 

```
... some hits for PHPStan:

- please install PHP7.1 as additional version, so that we can use the current version of PHPStan
- use ```/* @var $foo FooClass */``` right before the variable, if there is magic in the code and PhpStorm and PHPStan can't get the type of the variable
- use ```if ($foo instanceof FooClass) { }``` if the code comment did not help
- you can ignore magic stuff e.g. from the ActiveRecords (if you fill the object with something different via SQL) in the PHPStan configuration -> ignoreErrors  - message: - path:  (warning: "message" is reg-ex stuff, so you need to escape your string)
```

#### 2. Different configs or different settings in one config?!
The next thing I added were two versions of the PHPStan config (phpstan_legacy.neon && phpstan.neon) and on a commit (something like this git hook: https://gist.github.com/voku/87f6124734e0b2eab07aeaaf7f8db3b7) I check for new files and will use the non-legacy version then. Step by step we try to reduce the differences of the config files by fixing the legacy part.

=> So maybe not only something like a auto-ignore current problems as a starting point (https://github.com/phpstan/phpstan/issues/1689) it a good idea for legacy applications, but also different settings per file state (new, changed, non-changed...) or per directory (src/modules/shop,src/modules/orders,...).

#### 3. PHPStan & PhpStorm != :heart: 
"PHP-CS-Fixer" and "PHP_codesniffer" both have a good integration into PhpStorm (+ auto-configuration via composer.json) and everybody in the team can see the results in the IDE itself, this feature would be also very nice for PHPStan. I already saw you comment here https://youtrack.jetbrains.com/issue/WI-35063 

=> I also think that this will increase the visibility of PHPStan in the PHP world. :rocket:

#### 4. Run single "Rules"
For "PHP-CS-Fixer" and "PHP_codesniffer" I quite often run some single sniffs / fixes in the last time, so I can transform many many files at once for example to use "final" by default, to force correct usage of a translation class or to prevent to execute SQL inside of views or controllers. 

=> So if I could add custom "Rules" with the ".phar"-setup, then it would be good if I could run this without any other "Rules".

---

Finally, thanks for your work and sorry for the long text... :)

PS2: maybe you can also ask https://twitter.com/kalessil | @kalessil for some experiences with making money with a OSS project and/or how to sell a non OSS version like a "Pro" version
